javascript/pkce_oauth/server/index.js

```diff
@@ -0,0 +1,174 @@
+const clientSecret = process.env.client_secret;
+const redirectUri = process.env.redirect_uri;
+const frontEndLocation = process.env.front_end_location;
+const tokenExchangeEndpoint = process.env.token_exchange_endpoint;
+
+const stateCache = {};
+const tokenCache = {};
+
+const app = express();
+app.use(cookieParser());
+
+// Base endpoint. This should return our html.
+app.get('/', (req, res) => {
+    console.log('In /');
+
+    request(frontEndLocation).pipe(res);
+});
+
+// First step.
+// We need to generate state and code verifiers for this client.
+// We then return the info the client needs to hit the auth server.
+app.get('/first-step-auth', (req, res) => {
+    console.log('In first-step-auth');
+
+    let result = {};
+
+    let state = getRandomToken();
+    let codeVerifier = getRandomToken();
+    let codeChallenge = sha256(codeVerifier);
+
+    // Save that the state and codeVerifier were given to the same person
+    stateCache[state] = codeVerifier;
+
+    // Supply the client with the OAuth data
+    result.codeChallenge = codeChallenge;
+    result.challengeMethod = challengeMethod;
+    result.clientId = clientId;
+    result.redirectUri = redirectUri;
+
+    // Store state as a cookie on the client
+    res.cookie("state", state);
+
+    res.json(result);
+});
+
+// Second step.
+// We need to get the state and code from the client.
+// We will verify the state is valid, and send the code to the auth server.
+// We then store the refresh token locally and return the access token to the client.
+app.get('/second-step-auth', (req, res) => {
+    console.log('In second-step-auth');
+
+    let state = req.query.state;
+
+    // Confirm the given state matches the state in the user's cookies
+    if (state !== req.cookies.state) {
+        res.status(401);
+        res.send();
+        return;
+    }
+
+    // The user doesn't need this state anymore
+    res.clearCookie("state");
+
+    let code = req.query.code;
+    let codeVerifier = stateCache[state];
+
+    const requestBody = {
+        grant_type: authorizationGrantType,
+        client_id: clientId,
+        client_secret: clientSecret,
+        code_verifier: codeVerifier,
+        redirect_uri: redirectUri,
+        code: code,
+    };
+
+    // Lets ask the auth server for an access and refresh token
+    request.post(tokenExchangeEndpoint, {
+        form: requestBody
+    }, (error, responseObj, responseBody) => {
+        console.log('In response from asana');
+
+        if (error) {
+            console.error(error);
+            res.status(401);
+            res.send();
+            return;
+        }
+
+        handleNewToken(JSON.parse(responseBody), res);
+    });
+});
+
+// This allows the front end app to request a new token if their current token is not expired
+app.get('/refresh-token', (req, res) => {
+    console.log('In refresh-token');
+
+    let access_token = req.cookies.access_token;
+    let refreshTokenObj = tokenCache[access_token];
```

Comment:
Is the goal here to use only in-memory state? Storing a map from access token to refresh token feels weird compared to the typical mapping from user to refresh token. Maybe make a comment that this is not how tokens should be stored.

Reply:
Yeah, because we're using a google function and trying to stay in the free tier, we're limited to in memory. 

The audience is for implicit grant deprecatees. I wanted to leave the refresh token out entirely, but because we're returning it I felt like I needed to surface it somehow.